Attempt to make Locked Vehicles Opt-In

Lazejun opened the pull request:
## What type of PR is this.
1. [ ] Bug
2. [x] Change
3. [x] Enhancement

### What have you changed and why?
Information:
While digging through the Files i found out that the only time we apply "locks" is when Non Commanders buy vehicles.
This is done by setting the Variable OwnerX with the Players SteamID.
Unlocking before this change was by removing OwernX's SteamID.

Added a locked variable to be added on Vehicle Placement and when Vehicles are bought by Non Commanders.
Commander's bought vehicles stay unaffected, they are always open and cannot be locked.

They start with locked false, meaning they are unlocked, only the Owner of the Vehicle can lock and Unlock it.
I wanted to remain the previous behaiviour of the Mechanic so Members Ignore locks, Owner ignores locks, Group Members of the Owner will ignore locks.

Updated the Hints to fit the changes.

Updated the Joke if you try to unlock a Person.

### Please specify which Issue this PR Resolves.
closes #XXXX

### Please verify the following and ensure all checks are completed.

1. [ ] Have you loaded the mission in LAN host?
2. [ ] Have you loaded the mission on a dedicated server?

### Is further testing or are further changes required?
1. [ ] No
2. [ ] Yes (Please provide further detail below.)

### How can the changes be tested?
Steps: 
Can be tested with Bought and Ungaraged Vehicles.
As a Member you can always enter the Vehicle,
As the Owner you can always enter the Vehicle,
As a Non Member you can only enter the Vehicle if its unlocked.
```sqf
//Test by Not being the Owner, you cannot Lock the Vehicle anymore.
_veh = cursorObject;
_veh setVariable ["ownerX","1234",true];
//Test by not being a Member anymore, you cannot ignore the lock, you cannot enter the vehicle if its locked.
membersx = [];
```
********************************************************
Notes:
The only Issue i could see is Sell Vehicle, as that actually checks and uses OwnerX
https://github.com/official-antistasi-community/A3-Antistasi/blob/afd998fc28ec329333929e37ba11dcac1111063f/A3A/addons/core/functions/Base/fn_sellVehicle.sqf#L47

jaj22 commented on A3A/addons/core/functions/Base/fn_unlockVehicle.sqf:
```diff
@@ -4,16 +4,22 @@ _veh = cursorObject;
+if (_ownerX != getPlayerUID player) exitWith {["Unlock Vehicle", "You cannot unlock/lock vehicles which you do not own."] call A3A_fnc_customHint;};
 
-["Unlock Vehicle", "Vehicle Unlocked."] call A3A_fnc_customHint;
+_locked = _veh getVariable "locked";
```
Can we trust that the "locked" variable exists only if "ownerX" does? If not then it's safer to use defaults, eg. `_locked = _veh getVariable ["locked", false]`.

jaj22 commented on A3A/addons/core/functions/Base/fn_unlockVehicle.sqf:
```diff
@@ -4,16 +4,22 @@ _veh = cursorObject;
 
-["Unlock Vehicle", "Vehicle Unlocked."] call A3A_fnc_customHint;
+_locked = _veh getVariable "locked";
+if (_locked isequalto false) then {
+    _veh setVariable ["locked",true];
+    ["Unlock Vehicle", "Vehicle locked."] call A3A_fnc_customHint;
+} else {
+    _veh setVariable ["locked",false];
+    ["Unlock Vehicle", "Vehicle unlocked."] call A3A_fnc_customHint;	
+};
```
The locked variable definitely needs to be global here though, so `_veh setVariable ["locked", whatever, true]`.

isEqualTo on booleans is unnecessary because you can just do `if (_locked)` or `if (!_locked)`. Arguably more readable with the explicit comparison but Hakon will take the piss.

jaj22 commented on A3A/addons/core/functions/init/fn_initClient.sqf:
```diff
@@ -245,13 +245,14 @@ player addEventHandler ["GetInMan", {
-				if (_owner isEqualType "") then {
-					if ({getPlayerUID _x == _owner} count (units group player) == 0) then {
-						["Warning", "You cannot board other player vehicle if you are not in the same group."] call A3A_fnc_customHint;
+			_locked = _veh getVariable "locked";
+			if (_locked isequalto true) then {
```
Similarly this is an actual nil comparison, which tends to silently fail but isn't good practice. Better to use nil/true + default false on getVariable.

HakonRydland replied:
sqf is not statically typed, we cant be sure that a outside variable will be of a specific type so better to have isEqualTo which is a type safe comparison

jaj22 replied:
If there's a risk that it's a variable from outside (which there is, I guess), we should be renaming it to A3A_locked.

jaj22 commented on A3A/addons/core/functions/init/fn_initClient.sqf:
```diff
@@ -245,13 +245,14 @@ player addEventHandler ["GetInMan", {
-						["Warning", "You cannot board other player vehicle if you are not in the same group."] call A3A_fnc_customHint;
+			_locked = _veh getVariable "locked";
+			if (_locked isequalto true) then {
+				if (!isNil "_owner") then {
```
This one is probably skippable though, because it shouldn't be possible to lock a vehicle without setting an owner first, unless I missed something.

Lazejun replied:
Its realy weird Spaghetti, i was mostly suprised that it worked the way i wanted it to.

jaj22 commented on A3A/addons/core/functions/init/fn_initClient.sqf:
```diff
@@ -245,13 +245,14 @@ player addEventHandler ["GetInMan", {
+			_locked = _veh getVariable "locked";
+			if (_locked isequalto true) then {
+				if (!isNil "_owner") then {
+					if (_owner isEqualType "") then {
```
Is it ever not a string?

jaj22 commented on A3A/addons/Garage/Public/config.inc:
```diff
@@ -61,6 +61,8 @@ HR_GRG_fnc_Hint = {
 HR_GRG_fnc_vehInit = {
     if (_this isKindOf "StaticWeapon") then {staticsToSave pushBack _this; publicVariable "staticsToSave"};
+    _this setVariable ["ownerX",getPlayerUID player,true];
+    _this setVariable ["locked",false];
```
Needs to be a global set, `_this setVariable ["locked", false, true]`, or preferably just switch to nil/true and skip it entirely.

Lazejun commented:
Ok the Variable is now Global,
Renamed locked to A3A_locked
Simplified the checks for A3A_locked as you suggested
I removed the dead checks in InitClient, as it seems to work without them.
Still the same wanted behaiviour, Members ignore Locks, Owners Ignore locks, Members of Owners Group would ignore locks too (I can't test that Alone).

jaj22 commented on A3A/addons/core/functions/init/fn_initClient.sqf:
```diff
@@ -245,17 +245,14 @@ player addEventHandler ["GetInMan", {
 	_veh = _this select 2;
 	_exit = false;
-	if (isMultiplayer) then {
 		if !([player] call A3A_fnc_isMember) then {
```
Code from here is one level too deep.

jaj22 commented on A3A/addons/core/functions/init/fn_initClient.sqf:
```diff
@@ -245,17 +245,14 @@ player addEventHandler ["GetInMan", {
 	_exit = false;
-	if (isMultiplayer) then {
 		if !([player] call A3A_fnc_isMember) then {
 			_owner = _veh getVariable "ownerX";
```
Should move this inside the lock check now that's it's not used outside.

jaj22 commented on A3A/addons/core/functions/init/fn_initClient.sqf:
```diff
@@ -245,17 +245,14 @@ player addEventHandler ["GetInMan", {
-						_exit = true;
-					};
-				};
+			_A3A_locked = _veh getVariable "A3A_locked";
+			if (_A3A_locked) then {
```
Similar issues here, although in this case it'll probably work because there's no else statement. Better practice is still:
```
private _locked = _veh getVariable ["A3A_locked", false];
if (_locked) then {
```

Or the nil/true version:
`if (isNil { _veh getVariable "A3A_locked"}) then {`

Lazejun commented:
🥳  We shall try again
